### PR TITLE
(WIP) networking: Add support for OpenVPN

### PR DESCRIPTION
**Pull request by subhoghoshX**
image

### CODE REVIEW

File: pkg/networkmanager/dialogs-common.jsx [hunk 1]
```diff
@@ -234,6 +236,8 @@ export const NetworkAction = ({ buttonText, iface, connectionSettings, type }) =
             dlg = <BridgeDialog {...properties} />;
         else if (type == 'wg')
             dlg = <WireGuardDialog {...properties} />;
+        else if (type == 'openvpn')
+            dlg = <OpenVPNDialog {...properties} />;
```
Comment (cockpituous) [inline pkg/networkmanager/dialogs-common.jsx:240]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/interfaces.js [hunk 2]
```diff
@@ -733,8 +733,12 @@ export function NetworkManagerModel() {
                     }
                 };
             }));
-        } else {
+        } else
             delete result.wireguard;
+
+        if (settings.vpn) {
+            set("vpn", "service-type", "s", settings.vpn['service-type']);
+            set("vpn", "data", "a{ss}", settings.vpn.data);
```
Comment (cockpituous) [inline pkg/networkmanager/interfaces.js:741]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import cockpit from 'cockpit';
+import { ModelContext } from './model-context';
+import { useDialogs } from 'dialogs.jsx';
+
+const _ = cockpit.gettext;
+
+export function OpenVPNDialog({ settings, connection, dev }) {
+    const Dialogs = useDialogs();
+    const idPrefix = "network-openvpn-settings";
+    const model = useContext(ModelContext);
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:15]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    const [iface, setIface] = useState(settings.connection.interface_name);
+    const [remote, setRemote] = useState("");
+    const [caCertName, setCaCertName] = useState("");
+    const [caCertVal, setCaCertVal] = useState("");
+    const [userCertName, setUserCertName] = useState("");
+    const [userCertVal, setUserCertVal] = useState("");
+    const [userKeyName, setUserKeyName] = useState("");
+    const [userPrivateKey, setUserPrivateKey] = useState("");
+    const [dialogError, setDialogError] = useState("");
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:25]: These 9 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        const userCertPath = `${user.home}/.cert/cert-${userCertName}`;
+        const userKeyPath = `${user.home}/.cert/key-${userKeyName}.key`;
+        try {
+            await cockpit.script(`mkdir -p ${user.home}/.cert`);
+            await cockpit.script(`touch ${caPath} ${userCertPath} ${userKeyPath}`);
+            await cockpit.file(caPath).replace(caCertVal);
+            await cockpit.file(userCertPath).replace(userCertVal);
+            await cockpit.file(userKeyPath).replace(userPrivateKey);
+        } catch (err) {
+            console.log("an error occured => ", err);
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:39]: These 13 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        function createSettingsObject() {
+            return {
+                ...settings,
+                connection: {
+                    ...settings.connection,
+                    type: 'vpn',
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:47]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                },
+                vpn: {
+                    data: {
+                        remote,
+                        ca: caPath,
+                        cert: userCertPath,
+                        key: userKeyPath,
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:54]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                        // hardcoding the bellow properties until we have a design for advanced dialog
+                        'cert-pass-flags': '0',
+                        'connection-type': 'tls',
+                        dev: 'tun',
+                        'push-peer-info': 'yes',
+                        'remote-cert-tls': 'server'
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:60]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                    },
+                    'service-type': 'org.freedesktop.NetworkManager.openvpn'
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:62]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            };
+        }
+
+        dialogSave({
+            connection,
+            dev,
+            model,
+            settings: createSettingsObject(),
+            onClose: Dialogs.close,
+            setDialogError,
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:73]: These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        });
+    }
+
+    return (
+        <NetworkModal
+            title={!connection ? _("Add OpenVPN") : _("Edit OpenVPN settings")}
+            isCreateDialog={!connection}
+            onSubmit={onSubmit}
+            dialogError={dialogError}
+            idPrefix={idPrefix}
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:83]: These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/openvpn.jsx [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            <FormGroup label={_("User certificate")}>
+                <FileUpload id={idPrefix + '-user-cert'} filename={userCertName} onFileInputChange={(_, file) => setUserCertName(file.name)} type='text' onDataChange={(_, val) => setUserCertVal(val)} hideDefaultPreview />
+            </FormGroup>
+            <FormGroup label={_("User private key")}>
+                <FileUpload id={idPrefix + '-user-key'} filename={userKeyName} onFileInputChange={(_, file) => setUserKeyName(file.name)} type='text' onDataChange={(_, val) => setUserPrivateKey(val)} hideDefaultPreview />
+            </FormGroup>
+            <FormGroup label={_("Test")}>
+                <input type='file' id={idPrefix + '-test-upload'} />
+            </FormGroup>
+        </NetworkModal>
```
Comment (cockpituous) [inline pkg/networkmanager/openvpn.jsx:101]: These 17 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20230921-075022-79cb9fd2-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by subhoghoshX** [comment]
Greatly simplified config/cert generation, fetching to host machine and file upload tests.

**Comment by subhoghoshX** [comment]
353 ms to 140 ms = 60.4% reduction
1050 ms to 225 ms (in fast 3G) = 78.6% reduction

I guess the refactoring was worth it...

**Comment by subhoghoshX** [comment]
Generating Deffie-Hellman params on the CI can be an interesting [flake](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20231029-160315-b7795355-fedora-38-networking/log.html#40) as generating the (prime) numbers can take anywhere from less than a second to more than a minute...

**Review by martinpitt — CHANGES_REQUESTED** [review]
I had a first coarse look at the import and config file structure. I'm rather relieved to see that there is nothing fundamentally weird -- just some rather shallow and easy cleanups (see threads below). I did not look into the design aspect of this, that needs to wait until Garrett is back. I figure he has some ideas/opinions about the dialog and the optional import, and how to "merge" it with Wireguard to have a single "Add VPN" button. But let's get the technical underpinnings right in the meantime.

Thanks, this is great work!

**Comment by subhoghoshX** [comment]
I've seen this flake https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20231211-103844-f3eb448f-fedora-38-networking/log.html#40 couple times by now, but it doesn't happen locally. I'll dig into it.

**Comment by subhoghoshX** [comment]
I've seen this flake https://cockpit-logs.us-east-1.linodeobjects.com/pull-19363-20231211-103844-f3eb448f-fedora-38-networking/log.html#40 couple times by now, but it doesn't happen locally. I'll dig into it.

**Comment by subhoghoshX** [comment]
> I had a first coarse look at the import and config file structure. I'm rather relieved to see that there is nothing fundamentally weird...

Well the most fundamentally weird thing is still that temporarily importing a connection with `nmcli`, accessing a `.nmconnection` file (which you questioned whether it's a defined API of NM, it seems to be, but with the recent netplan changes in ubuntu, in #19724, I can't guarantee anymore), creating a config file and cleaning up the mess. I don't like the whole approach overall.

On Monday I found this [KDE bug report](https://bugs.kde.org/show_bug.cgi?id=396530) and this [commit closing it](https://invent.kde.org/plasma/plasma-nm/-/commit/9b22aa548f02ef43f7275ab2f9ffa48bde8a1ba8) that uses a shared object exposed by NM. It's in C++, over the weekend I'll see if can do something similar in Python.

**Comment by martinpitt** [comment]
This is rather stale, needs a lot of work, and there's currently nobody who could pick that up. So I'm closing this now. We can reopen it if/when there is someone interested in working on this again. Thanks!

**Pull request closed without merging**